src/toast/ops/conviqt.py

```diff
@@ -0,0 +1,693 @@
+        4.0 * u.arcmin,
+        allow_none=False,
+        help="Width of a symmetric gaussian beam already present in the skyfile "
+        "(will be deconvolved away).",
+    )
+
+    sky_file_dict = Dict(
+        None,
+        allow_none=True,
+        help="Dictionary of files containing the sky a_lm expansions. An entry for "
+        "each detector name must be present. If provided, supersedes `sky_file`.",
+    )
+
+    sky_file = Unicode(
+        None,
+        allow_none=True,
+        help="File containing the sky a_lm expansion.  Tag {detector} will be "
+        "replaced with the detector name",
+    )
+
+    beam_file_dict = Dict(
+        None,
+        allow_none=True,
+        help="Dictionary of files containing the beam a_lm expansions. An entry for "
+        "each detector name must be present. If provided, supersedes `beam_file`.",
+    )
+
+    beam_file = Unicode(
+        None,
+        allow_none=True,
+        help="File containing the beam a_lm expansion.  Tag {detector} will be "
+        "replaced with the detector name.",
+    )
+
+    @traitlets.validate("shared_flag_mask")
+    def _check_shared_flag_mask(self, proposal):
+        check = proposal["value"]
+        if check < 0:
+            raise traitlets.TraitError("Shared flag mask should be a positive integer")
+        return check
+
+    @traitlets.validate("det_flag_mask")
+    def _check_det_flag_mask(self, proposal):
+        check = proposal["value"]
+        if check < 0:
+            raise traitlets.TraitError("Det flag mask should be a positive integer")
+        return check
+
+    def __init__(self, **kwargs):
+        super().__init__(**kwargs)
+        return
+
+    @property
+    def available(self):
+        """Return True if libconviqt is found in the library search path."""
+        return conviqt is not None and conviqt.available
+
+    hwp_angle = Unicode(
+        None, allow_none=True, help="Observation shared key for HWP angle"
+    )
+
+    @traitlets.validate("hwp_angle")
+    def _check_for_hwp_angle(self, proposal):
+        # Check that there is no rotating HWP
+        check = proposal["value"]
+        if check is not None:
+            raise RuntimeError(
+                "SimConviqt cannot handle a HWP. Please use SimWeightedConviqt"
+            )
+        return check
+
+    @function_timer
+    def _exec(self, data, detectors=None, **kwargs):
+        if not self.available:
+            raise RuntimeError("libconviqt is not available")
+
+        log = Logger.get()
+
+        timer = Timer()
+        timer.start()
+
+        # Expand detector pointing
+        if self.quats:
+            quats_name = self.quats
+            keep_quats = True
+        else:
+            quats_name = "quats"
+            keep_quats = False
+        self.detector_pointing.quats = quats_name
+        self.detector_pointing.apply(data, detectors=detectors)
+
+        all_detectors = self._get_all_detectors(data, detectors)
+
```

Comment:
Before going into the detector loop below, we should check that the output exists for all detectors here.  That way we don't need the call inside `calibrate_signal` and `cache`, where the `ensure()` function will be called for the full detector set, but once for each detector.
```
for ob in data.obs:
    ob_dets = ob.select_local_detectors(detectors)
    ob.detdata.ensure(self.det_data, detectors=ob_dets)
```

Reply:
Yup.